### PR TITLE
Implemented inline assembly

### PR DESCRIPTION
**Pull request by tillspeicher**

### CODE REVIEW

File: src/thorin/be/llvm/llvm.cpp [hunk 1]
```diff
@@ -888,6 +900,69 @@ llvm::Value* CodeGen::emit_lea(const LEA* lea) {
...
+    for (auto clob : inl_asm->clobbers())
+        constraints += "~" + clob + ",";
+    // clang always marks those registers as clobbered, so we will do as well
+    constraints += "~{dirflag},~{fpsr},~{flags}";
+
+    if (!llvm::InlineAsm::Verify(fn_type, constraints)) {
+        delete input_params;
+        delete param_types;
+        ELOG("Constraints and input and output types of inline assembly do not match at %", inl_asm->loc());
+        // TODO: nullptr ok?
```
Comment (richardmembarth) [inline src/thorin/be/llvm/llvm.cpp]: ELOG calls std::abort - the 'return' can be omitted.

**Comment by leissa** [comment]
Hi,

the patch looks nice :)

Some minor technical remarks: 

* I think `Asm` would be the perfect name for this node, but since `asm` is already a C++ keyword I suggest `Assembly` - this is less cryptic than `inl_asm` and then we can use `assembly` as variable name
* use `Types` instead of `std::vector<const Type*>&`
* use `Defs` instead of `std::vector<const Def*>&`
* use `ArrayRef<std::string>` instead of `std::vector<std::string>` in the constructor and `Array<std::string>` instead of `std::string<std::string>` as member.
* simply assume that the input tuple already has 'mem' as first element. You can offer a second factory method in `World` where `mem` can be specified as extra input item.
* same story for the output types.
* remove the default constructor arguments - they are useless because the user can only use the factory method
* to be consistent, the `const Location& loc` should be the last argument.
* I think it would be better to have an appropriate `enum` with all flags instead of passing a ton of boolean flags - it's also easier to add further flags if this is desired.

To sum it up, the constructor should look like this:
```C++
Assembly(const Type* type, Defs input, std::string asm_template, ArrayRef<std::string> output_constraints, ArrayRef<std::string> input_constraints, ArrayRef<std::string> clobbers, Flags flags, const Location& loc)
```
Same story for the factory method + a second method which looks like this:
```C++
assembly(Types types, const Def* mem, Defs input, std::string asm_template, ArrayRef<std::string> output_constraints, ArrayRef<std::string> input_constraints, ArrayRef<std::string> clobbers, Flags flags, const Location& loc)
```

File: src/thorin/primop.h [hunk 2]
```diff
@@ -583,16 +590,16 @@ class Assembly : public MemOp {
...
+    bool is_alignstack() const { return flags_ & Assembly::Flags::IS_ALIGNSTACK; }
+    bool is_inteldialect() const { return flags_ & Assembly::Flags::IS_INTELDIALECT; }
 
 private:
     virtual const Def* vrebuild(World& to, Defs ops, const Type* type) const override;
 
     std::string template_;
     Array<std::string> output_constraints_, input_constraints_, clobbers_;
-    bool has_sideeffects_, is_alignstack_, is_inteldialect_;
+    Assembly::Flags flags_;
```
Comment (leissa) [inline src/thorin/primop.h]: You can elide `Assembly::` within `Assembly`

File: src/thorin/primop.h [hunk 3]
```diff
@@ -573,8 +573,15 @@ class Enter : public MemOp {
 };
 
 class Assembly : public MemOp {
+public:
+    enum Flags {
+        HAS_SIDEEFFECTS = (1u << 0),
+        IS_ALIGNSTACK   = (1u << 1),
+        IS_INTELDIALECT = (1u << 2),
+    };
+
```
Comment (leissa) [inline src/thorin/primop.h]: Convention is:
```C++
enum Flags {
    HasSideeffects = 1 << 0,
    IsAlignstack   = 1 << 1,
    IsInteldialect = 1 << 2,
};
```
All-capital looks like preprocessor stuff.

File: src/thorin/be/llvm/llvm.cpp [hunk 4]
```diff
@@ -888,6 +900,74 @@ llvm::Value* CodeGen::emit_lea(const LEA* lea) {
...
+            res_type = convert_tuple(out_type->args().begin(),
+                    out_type->args().end(), out_type->num_args());
+            break;
+    }
+
+    // TODO: better way to get the input params and types?
+    auto ops = assembly->ops();
+    auto op_size = ops.size() - 1;       // omit the mem type
+    auto input_params = new llvm::Value *[op_size];
+    auto param_types = new llvm::Type *[op_size];
```
Comment (leissa) [inline src/thorin/be/llvm/llvm.cpp]: don't use `new`. Use `thorin::Array` instead. Then, you can also get rid of the `deletes`.

Comment (tillspeicher) [inline src/thorin/be/llvm/llvm.cpp]: I can't use `thorin::Array` because I need to pass `input_params` and `param_types` to LLVM functions which don't accept it. I will use `std::vector` instead.

Comment (leissa) [inline src/thorin/be/llvm/llvm.cpp]: you can use `llvm_ref` to do the magic. See other uses in this file. Declaration is in `llvm.h`.

**Comment by tillspeicher** [comment]
I addressed all the issues, it should be ready to merge now.

**Comment by leissa** [comment]
I've also polished a bit and fixed some potential problems. Can you plz have a look at my chagnes that I didn't break anything?
Note that I have removed the static method `get_flags`. Simply use, for instance: `HasSideEffects | IsAlignStack`. Maybe, you want to reuse in impala the same enum. So you don't have to convert stuff.

**Comment by tillspeicher** [comment]
I'm just updating Impala based on the changes you made.
Using the `Flags` enum there would be nice, but then we need to include `primop.h` in `ast.h` which I guess we don't want.
In this case we should move the enum somewhere else and include that.
Is `enums.h` the right place for that?

**Comment by leissa** [comment]
Simply include `primop.h`. That's fine.
